### ISSUE TITLE
MParticle.start taking too much time

### ISSUE THREAD

**Issue by jcesarmobile** [mParticle/mparticle-android-sdk#79]
I have an app that uses a WebView and calling `MParticle.start` is taking soo much time to run on some devices.

My code looks like this:

```
MParticle.start(options);
MParticle.getInstance().registerWebView(myWebView, "myBridgeName");
myWebView.loadUrl("path/to/index.html");
```

Since start is taking so long (around 6-10 seconds), my app can't load the content for those 6-10 seconds (plus other 2 seconds that my app takes already)

If I put MParticle code in an Executor, then loadUrl is called before MParticle is initialized, which I think can cause problems, at least [the docs](https://docs.mparticle.com/developers/sdk/web/native-webviews/#android-sdk) show that should be called after.

> //note that this *must* be before you load the page

My device is a Motorola Moto G with Android 5, but it has also been reported on some Samsung Galaxy S9 with Android 10, which aren't old nor slow.

SDK version 5.25.0, but it was slow in older versions too
Updated to 5.33.1 and still reproduces.

**Comment by alexs-mparticle** [comment]
@jcesarmobile Thanks for opening this issue. I'll link up with our android team to investigate and get back to you.

**Comment by alexs-mparticle** [comment]
Tracked internally as 76317.

**Comment by alexs-mparticle** [comment]
@jcesarmobile This issue has been routed to our android team's backlog and will be prioritized accordingly. Please be aware that we are in a release blackout until January, so we cannot release a fix until at least then.

If we do find a resolution sooner, we will update this issue accordingly.

**Comment by jcesarmobile** [comment]
Thanks for the update

**Comment by fparhat-rbi** [comment]
We are hitting this issue with our applications as well.
There are some key environment information information to repro:
- Point to a mParticle workspace with significantly large amount of events + attributes + filters (50k+)
- Install 3+ kits (Appboy, Google Analytics, Branch)
- Deploy app to an Android device such as a Galaxy S9 (note this issue does not repro running on Android emulator)

**Comment by fparhat-rbi** [comment]
Update:
- We were able to reproduce the issue on a bare react native app (not using expo) where the only addition was the react-native-mparticle plugin + Android SDKs + pointing to our PROD workspace

**Comment by einsteinx2** [comment]
@fparhat-rbi Thank you, these repro details are very helpful! I'll add this information to our internal ticket.

**Comment by willpassidomo** [comment]
@jcesarmobile @fparhat-rbi I've been doing some digging and found something that could potentially cause some drag on startup when the remote config we fetch is very large. I'm skeptical that it could cause as extreme of a hang as you guys are experiencing, but it definitely seems inefficient. 

Would either of you be able to confirm whether this lag happens only on a fresh install (first start, clean app storage), only on starts _after_ the fresh install or both?

**Comment by fparhat-rbi** [comment]
Hi Will,
The lag occurs on all cases mentioned above: after a fresh install and on every cold launch (i.e killing the app and launching it again).

In our case, removing the 3 kit integrations we use (appboy, google-analytics-firebase, and branch) dramatically reduces the time it takes to launch the applicatio.

**Comment by willpassidomo** [comment]
@fparhat-rbi I've been able to replicate a startup time close to 500ms and have a fix ready for that issue, but I haven't been able to reproduce anything similar to the 1-5 second range you guys mentioned. Would it be possible for you to share a minimal project that  reproduces the issue? 

I want to make sure I'm not missing something in the interplay between the Android SDK, React wrapper SDK and the JS SDK, presumably running in the webview (although it sounds like you don't need that to reproduce)

**Comment by willpassidomo** [comment]
@jcesarmobile @fparhat-rbi We released a fix for this just now in version `5.36.0`. Our testing revealed that very large configuration responses, generally caused by a large number of kit filtering/projection rules were causing excessive load times for a `SharedPreferences` instance we opened during the first `MParticle.start()` invacation. In our testing, this decreased the SDK startup times for a 1 mb configuration (~100,000 filters) by around 85%.

Our testing uncovered a few other potential problem spots in our code when large kit configurations are present and we will be addressing them in our upcoming releases. Please feel free to write back here or send me a message if you are still experiencing issues after upgrading to the latest version!
